File: packages/sdk-kotlin/src/main/kotlin/dev/rovenue/sdk/paywallui/NodeVisibility.kt
package dev.rovenue.sdk.paywallui

import android.graphics.Rect
import android.view.View
import android.view.ViewTreeObserver
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.LifecycleEventObserver
import androidx.lifecycle.ProcessLifecycleOwner

// =============================================================
// ONE visibility rule for every time-driven node.
//
// Every timer, player and animation in a paywall answers the same question:
// is this node on screen RIGHT NOW? Two independent signals compose, and a
// node is only allowed to burn time when BOTH say yes:
//
//   1. the node intersects the viewport   (geometry — [isNodeOnScreen])
//   2. the app is in the foreground       (process lifecycle)
//
// Signal 2 already existed on Android (the carousel's `ProcessLifecycleOwner`
// observer); signal 1 did not. This file is where both now live, ONCE, so a
// new time-driven node type inherits the rule instead of re-deriving it.
//
// CROSS-PLATFORM CONTRACT. The web renderer uses `IntersectionObserver` +
// `visibilitychange`; the SwiftUI renderer uses a `GeometryReader` in a named
// coordinate space + scene phase. Android's equivalent of the first is
// `View.getLocalVisibleRect` sampled from a
// `ViewTreeObserver.OnScrollChangedListener` — see [NodeVisibilityDetector].
// All three FAIL OPEN when they cannot tell (see [isNodeOnScreen]).
// =============================================================

/** Seconds → millis, for the authored auto-advance interval. */
internal const val MILLIS_PER_SECOND = 1000.0

/**
 * The fewest pages a carousel needs before auto-advance means anything:
 * one page (or none) has nowhere to advance to.
 */
internal const val CAROUSEL_MIN_AUTO_ADVANCE_PAGES = 2

/**
 * The smallest authored auto-advance interval that is actually an interval.
 * At or below this, auto-advance is OFF — `postDelayed(0)` would spin the
 * pager as fast as the main looper can deliver.
 */
private const val MIN_AUTO_ADVANCE_SECONDS = 0.0

/**
 * Geometry half of the rule: does [visibleRect] — what
 * `View.getLocalVisibleRect` reported for a view of [viewWidth] x
 * [viewHeight], or `null` when it reported nothing visible at all — mean this
 * node is on screen?
 *
 * ANY intersection counts. A node one pixel into the viewport is a node the
 * user can see a timer on, and there is deliberately no "at least N percent"
 * threshold: the three renderers would have to agree on the fraction, and web
 * and iOS do not apply one either.
 *
 * FAIL OPEN, and note the ORDER — the unmeasured case is decided FIRST. Before
 * layout a view has no size and no visible rect, and reading that as
 * "off screen" would stop every timer at attach and leave a paywall full of
 * frozen countdowns. A stopped clock is worse than a running one, so an
 * unmeasured view counts as on screen (matching web, where a node with no
 * `IntersectionObserver` entry yet is treated as visible, and iOS, where a
 * `.zero` viewport counts as visible). Only once the view HAS a size does a
 * missing visible rect actually mean "scrolled away".
 *
 * NOT OBSERVABLE OFF-DEVICE: under the mockable `android.jar` every `Rect`
 * accessor is inert (fields stay `0`, `isEmpty()` returns the default
 * `false`), so the unit tests pin the unmeasured and null-rect branches only.
 * The `isEmpty()` branch — a degenerate rect from a zero-area intersection —
 * is real device behaviour that a JVM test cannot reach.
 */
internal fun isNodeOnScreen(visibleRect: Rect?, viewWidth: Int, viewHeight: Int): Boolean {
    if (viewWidth <= UNMEASURED_VIEW_DIMENSION_PX || viewHeight <= UNMEASURED_VIEW_DIMENSION_PX) return true
    val rect = visibleRect ?: return false
    return !rect.isEmpty()
}

/**
 * The whole rule: a node may run time-driven work only while it is
 * [onScreen] AND the app is [appForegrounded].
 *
 * A backgrounded app's main looper keeps delivering messages, so neither
 * signal implies the other — an on-screen carousel in a backgrounded app
 * still advances pages and CHANGES THE PAGE THE USER COMES BACK TO unless the
 * foreground half is checked too.
 */
internal fun isNodeTimerActive(onScreen: Boolean, appForegrounded: Boolean): Boolean =
    onScreen && appForegrounded

/**
 * Whether a `countdown` row's once-a-second tick should be running:
 * [active] (the composed rule) and not already collapsed by
 * `onExpiry: "hide"`.
 *
 * [hiddenOnExpiry] is a latch that never un-latches, so it survives a
 * pause/resume cycle: coming back on screen must not resurrect the tick of a
 * countdown that has nothing left to draw.
 */
internal fun countdownTickShouldRun(active: Boolean, hiddenOnExpiry: Boolean): Boolean =
    active && !hiddenOnExpiry

/**
 * How long until a `carousel` should turn its next page, or `null` when it
 * should not be scheduled at all.
 *
 * `null` covers four separate reasons, which is exactly why this is one
 * function and not four scattered guards: the node is paused ([active] is
 * false), a `loop: false` carousel already latched [stoppedAtEnd], there is
 * no positive authored interval ([autoAdvanceSeconds] absent means OFF, not
 * "use a default"), or there is nowhere to advance to.
 *
 * [stoppedAtEnd] is checked independently of [active], so the stop latch
 * SURVIVES a pause/resume: pausing is not the same as reaching the last page,
 * and a carousel that stopped at its end must stay stopped when it scrolls
 * back into view.
 */
internal fun carouselAutoAdvanceDelayMillis(
    active: Boolean,
    stoppedAtEnd: Boolean,
    autoAdvanceSeconds: Double?,
    pageCount: Int,
): Long? {
    if (!active || stoppedAtEnd) return null
    if (pageCount < CAROUSEL_MIN_AUTO_ADVANCE_PAGES) return null
    val seconds = autoAdvanceSeconds ?: return null
    if (seconds <= MIN_AUTO_ADVANCE_SECONDS) return null
    return (seconds * MILLIS_PER_SECOND).toLong()
}

/**
 * The shared plumbing that keeps [isNodeTimerActive] up to date for one
 * [view], and calls [onActiveChanged] when the answer flips.
 *
 * Built ONCE per time-driven node and owned by it: [start] from
 * `onAttachedToWindow`, [stop] from `onDetachedFromWindow`. It supplies an
 * ADDITIONAL pause signal — it does not replace the owner's own handler
 * teardown, which must still happen on detach.
 *
 * EVERY LISTENER REGISTERED IS UNREGISTERED IN [stop]. `RovenuePaywallView`
 * rebuilds its whole view tree on every state change (a package tap is a
 * rebuild), and an attached view's `getViewTreeObserver()` is the WINDOW's
 * observer, shared by the entire hierarchy and outliving any single node — a
 * listener left behind both leaks the discarded view and re-runs on every
 * future scroll, once more per tap, for the life of the window. That exact
 * defect has already shipped in this renderer once.
 *
 * FAILS SOFT ON THE LIFECYCLE HALF. `ProcessLifecycleOwner.get()` THROWS when
 * `androidx.startup`'s `ProcessLifecycleInitializer` never ran — a host app is
 * free to strip it out of the manifest (`tools:node="remove"`), and apps that
 * manage their own App Startup do exactly that. That throw would land in
 * `onAttachedToWindow`, taking the host app down as the paywall appears. A
 * paywall must never crash a host app over an OPTIONAL nicety, and pausing
 * while backgrounded is exactly that — so a missing process lifecycle is
 * treated as "always foregrounded" and only the geometry half applies.
 *
 * DEVICE-ONLY: whether `OnScrollChangedListener` fires, and whether
 * `getLocalVisibleRect` returns real rects, is not observable under the
 * mockable `android.jar` (both are stubbed). The decision this class feeds is
 * pinned by unit tests; the plumbing that feeds it is a device smoke item.
 */
internal class NodeVisibilityDetector(
    private val view: View,
    private val onActiveChanged: (Boolean) -> Unit,
) {
    /** Reused across every sample so scroll-rate polling allocates nothing. */
    private val scratchRect = Rect()

    /** Resolved ONCE so [stop] removes the observer from the very lifecycle
     *  [start] added it to. `null` = this app has no process lifecycle. */
    private val processLifecycle: Lifecycle? =
        runCatching { ProcessLifecycleOwner.get().lifecycle }.getOrNull()

    /** With no process lifecycle there is nothing to tell us the app went
     *  away, so the foreground half is held permanently true rather than
     *  permanently false — the cost is the pre-fix behaviour, not a paywall
     *  whose timers never start. */
    private var appForegrounded: Boolean = processLifecycle == null

    /**
     * The current answer, and the reason a consumer can read it synchronously
     * (the carousel re-derives its delay on every page change, not only when
     * visibility flips). Starts `true` for the same fail-open reason
     * [isNodeOnScreen] does: before [start] there has been no sample.
     */
    var isActive: Boolean = true
        private set

    /** Last value handed to [onActiveChanged], or `null` if nothing has been
     *  published since the most recent [start]. */
    private var published: Boolean? = null

    private val foregroundObserver = LifecycleEventObserver { _, event ->
        when (event) {
            Lifecycle.Event.ON_START -> {
                appForegrounded = true
                publish()
            }

            Lifecycle.Event.ON_STOP -> {
                appForegrounded = false
                publish()
            }

            else -> Unit
        }
    }

    /** Android's `IntersectionObserver`: the window's scroll signal, which is
     *  what changes whether this node intersects the viewport. */
    private val scrollListener = ViewTreeObserver.OnScrollChangedListener { publish() }

    /** Scrolling is not the only way a node's position changes — the FIRST
     *  layout pass is what turns an unmeasured (fail-open) node into one whose
     *  real position is known, and no scroll need ever follow it. Without
     *  this, a node laid out below the fold would keep ticking until the user
     *  happened to scroll. */
    private val layoutListener = ViewTreeObserver.OnGlobalLayoutListener { publish() }

    fun start() {
        view.viewTreeObserver.takeIf { it.isAlive }?.let { observer ->
            observer.addOnScrollChangedListener(scrollListener)
            observer.addOnGlobalLayoutListener(layoutListener)
        }
        // Adding an observer to a `LifecycleRegistry` immediately replays the
        // owner's current state, so a foregrounded app delivers ON_START (and
        // therefore a publish) from inside this call, while a backgrounded one
        // correctly delivers nothing.
        processLifecycle?.addObserver(foregroundObserver)
        // ...which is why start() publishes explicitly as well: the
        // backgrounded and no-lifecycle cases must still get their first
        // answer. publish() de-duplicates, so the foregrounded case does not
        // fire twice.
        publish()
    }

    fun stop() {
        view.viewTreeObserver.takeIf { it.isAlive }?.let { observer ->
            observer.removeOnScrollChangedListener(scrollListener)
            observer.removeOnGlobalLayoutListener(layoutListener)
        }
        processLifecycle?.removeObserver(foregroundObserver)
        // Back to the fail-open starting point so a re-attach re-publishes
        // rather than being swallowed as "no change".
        published = null
        isActive = true
    }

    /**
     * Samples both signals and notifies ONLY on a change.
     *
     * The de-duplication is load-bearing, not an optimisation:
     * `OnScrollChangedListener` fires on every scrolled frame, and a consumer
     * that removes and re-posts its handler callback on each one would have
     * its timer permanently reset — a countdown that never ticks while the
     * user is scrolling, and a carousel that never advances.
     */
    private fun publish() {
        val next = isNodeTimerActive(
            onScreen = isNodeOnScreen(currentVisibleRect(), view.width, view.height),
            appForegrounded = appForegrounded,
        )
        isActive = next
        if (next == published) return
        published = next
        onActiveChanged(next)
    }

    /** `getLocalVisibleRect` returns false — and leaves [scratchRect]
     *  untouched — when no part of the view is visible, which is exactly the
     *  `null` [isNodeOnScreen] expects. */
    private fun currentVisibleRect(): Rect? =
        if (view.getLocalVisibleRect(scratchRect)) scratchRect else null
}

File: packages/sdk-swift/Sources/Rovenue/PaywallUI/NodeVisibility.swift
//
//  NodeVisibility.swift
//  ONE rule for every time-driven node: a `countdown`'s clock, a
//  `carousel`'s auto-advance — and anything the media wave adds — runs only
//  while the node is actually on screen AND the app is in front. Spec §5
//  rule 1, "off-screen means paused".
//
//  This is NOT the `visibility` gate in Visibility.swift, which answers a
//  different question (does this node render on this platform/app version at
//  all). This file answers "is the node the reader is looking at right now".
//
//  The web renderer gets this from `IntersectionObserver`
//  (packages/paywall-renderer's node-visibility hook). SwiftUI has no
//  equivalent, so it is assembled from two primitives:
//
//  - a `GeometryReader` background on the node, reporting the node's frame in
//    a NAMED coordinate space anchored on the paywall's scroll viewport
//    (`paywallViewportCoordinateSpaceName`, established in
//    `RovenuePaywallView.content(_:)`), compared against that viewport's own
//    frame, which the same place publishes through
//    `EnvironmentValues.paywallViewportFrame`;
//  - `scenePhase`, for the app-in-front half.
//
//  The pure DECISION (`isNodeOnScreen` / `isNodeRunning`) is deliberately
//  kept apart from that plumbing: the decision is unit-testable on any
//  platform and is pinned in PaywallRenderSupportTests, while whether SwiftUI
//  ever hands the modifier a real rect is only observable on a device and
//  stays a smoke item. Do not fold the two back together — a predicate buried
//  in a `ViewModifier` body is a rule nothing can check.
//
//  The governing rule, as everywhere else in this renderer, is FAIL OPEN.
//  Before first layout the viewport is `.zero` and the node's frame is
//  `.zero`; both resolve to ON SCREEN. A stopped clock is a worse failure
//  than a running one, and the web renderer documents the same default (with
//  no `IntersectionObserver` available, the node counts as visible).
//

import Foundation
import SwiftUI

/// The paywall's scroll-viewport coordinate space. Node frames are measured
/// in it, so a node's `minY` is its offset from the TOP OF THE VIEWPORT (it
/// goes negative as the reader scrolls past it), not from the top of the
/// scrolled content — which is what makes comparing against the viewport
/// rect meaningful at all.
let paywallViewportCoordinateSpaceName = "rovenue.paywall.viewport"

/// How much of a node must fall inside the viewport before it counts as on
/// screen, in points. One point: the rule is "any of it is showing", matching
/// the web hook's default `threshold: 0` (fire as soon as a single pixel
/// intersects) rather than an "at least half visible" rule, which would pause
/// a tall node the reader is already reading.
///
/// Named rather than inlined at the comparison — a bare `1` in a geometry
/// test reads as an off-by-one guard, not as a policy anyone chose.
let nodeOnScreenMinimumIntersectionPoints: CGFloat = 1.0

/// Everything the run/pause decision depends on, in one value, so the
/// decision stays a pure function of it and the modifier's only job is
/// keeping it current.
///
/// `unknown` is the pre-layout value every consumer starts from: no measured
/// node frame, no viewport, app assumed in front. It resolves to RUNNING —
/// see the fail-open note at the top of this file.
struct NodeVisibilityState: Equatable {
    var nodeFrame: CGRect
    var viewportFrame: CGRect
    var appIsForeground: Bool

    static let unknown = NodeVisibilityState(
        nodeFrame: .zero, viewportFrame: .zero, appIsForeground: true)
}

/// The pure geometry decision: does `nodeFrame` show inside `viewportFrame`?
/// Both rects are in the same (viewport-anchored) coordinate space.
///
/// Fails open on a viewport with no area — that is the state before first
/// layout, and also what a host that never establishes the paywall's
/// environment value leaves behind (an ordinary node rendered outside
/// `RovenuePaywallView`, say). Treating it as off-screen would stop every
/// timer at launch, which is exactly the bug this whole mechanism exists to
/// avoid causing.
///
/// A node with no area of its own also fails open, for the same reason: a
/// node that has not laid out yet reports `.zero`, and "we cannot tell yet"
/// must never read as "off screen". The `min(...)` is what does that — the
/// requirement is one point of intersection OR the node's whole extent,
/// whichever is smaller.
func isNodeOnScreen(nodeFrame: CGRect, viewportFrame: CGRect) -> Bool {
    guard viewportFrame.width > 0, viewportFrame.height > 0 else { return true }
    let intersection = nodeFrame.intersection(viewportFrame)
    // `.intersection` returns the NULL rect for disjoint rects — a sentinel
    // whose components are infinite, so it must be checked before any
    // arithmetic reads them.
    guard !intersection.isNull else { return false }
    let requiredWidth = min(nodeOnScreenMinimumIntersectionPoints, nodeFrame.width)
    let requiredHeight = min(nodeOnScreenMinimumIntersectionPoints, nodeFrame.height)
    return intersection.width >= requiredWidth && intersection.height >= requiredHeight
}

/// The whole rule: on screen AND the app in front. Every time-driven node
/// asks this one question — a second implementation of it anywhere is the
/// failure this file exists to prevent.
func isNodeRunning(_ state: NodeVisibilityState) -> Bool {
    state.appIsForeground
        && isNodeOnScreen(nodeFrame: state.nodeFrame, viewportFrame: state.viewportFrame)
}

/// Measures the node it is applied to and reports a fresh
/// `NodeVisibilityState` whenever any input changes: the node's frame (it
/// scrolled), the viewport's frame (rotation/resize), or the app's scene
/// phase.
///
/// The modifier NEVER decides anything itself; it hands the state to the
/// consumer, which combines it with its own preconditions (a carousel also
/// needs an interval, more than one page, and an unset stop latch). That
/// split is what lets each consumer's run/pause rule be a tested pure
/// function.
///
/// `appIsForeground` starts `true` and is only ever updated from a CHANGE of
/// `scenePhase`, never from its initial value. A SwiftUI view hosted from
/// UIKit without a scene publishing a phase would otherwise read a stale
/// non-active value and disable every timer outright; reacting to changes
/// only degrades to "always in front" there instead. This preserves the exact
/// behaviour `CarouselView` had before the two consumers moved onto this
/// modifier.
struct NodeVisibilityModifier: ViewModifier {
    let onChange: (NodeVisibilityState) -> Void

    @Environment(\.scenePhase) private var scenePhase
    @Environment(\.paywallViewportFrame) private var viewportFrame

    @State private var nodeFrame: CGRect = .zero
    @State private var appIsForeground = true

    func body(content: Content) -> some View {
        content
            // A background `GeometryReader` measures WITHOUT affecting the
            // node's own layout — wrapping the node in one instead would hand
            // it the parent's full proposed size and resize the paywall.
            .background(
                GeometryReader { proxy in
                    Color.clear.preference(
                        key: NodeFrameKey.self,
                        value: proxy.frame(in: .named(paywallViewportCoordinateSpaceName)))
                }
            )
            .onPreferenceChange(NodeFrameKey.self) { frame in
                nodeFrame = frame
                publish(nodeFrame: frame)
            }
            .onChange(of: viewportFrame) { newViewport in
                publish(viewportFrame: newViewport)
            }
            .onChange(of: scenePhase) { phase in
                let foreground = phase == .active
                appIsForeground = foreground
                publish(appIsForeground: foreground)
            }
            // The first report: a node that lays out entirely on screen and
            // never moves again produces no preference CHANGE after its
            // initial one, so consumers would otherwise sit on
            // `NodeVisibilityState.unknown` forever. Harmless that it does —
            // `unknown` runs — but the state the consumer holds should be the
            // real one from the start.
            .onAppear { publish() }
    }

    /// Each caller passes the value IT just changed and reads the others from
    /// storage, so nothing depends on a `@State` write being visible to a
    /// read in the same turn.
    private func publish(
        nodeFrame: CGRect? = nil, viewportFrame: CGRect? = nil, appIsForeground: Bool? = nil
    ) {
        onChange(NodeVisibilityState(
            nodeFrame: nodeFrame ?? self.nodeFrame,
            viewportFrame: viewportFrame ?? self.viewportFrame,
            appIsForeground: appIsForeground ?? self.appIsForeground))
    }
}

extension View {
    /// Applies `NodeVisibilityModifier`. Spelled as a `View` extension so a
    /// consumer reads `.nodeVisibility { ... }` alongside its other
    /// lifecycle modifiers.
    func nodeVisibility(_ onChange: @escaping (NodeVisibilityState) -> Void) -> some View {
        modifier(NodeVisibilityModifier(onChange: onChange))
    }
}

/// The node's own frame, in the paywall viewport's coordinate space.
private struct NodeFrameKey: PreferenceKey {
    /// `let`, not `var`, for the same reason as `StickyFooterHeightKey`: a
    /// mutable static on a `PreferenceKey` is shared global state nothing
    /// writes, and it trips Swift 6 strict concurrency.
    static let defaultValue: CGRect = .zero
    static func reduce(value: inout CGRect, nextValue: () -> CGRect) {
        value = nextValue()
    }
}

/// The paywall's scroll viewport, in its OWN coordinate space — i.e. origin
/// `.zero`, size the visible scroll area. Published by
/// `RovenuePaywallView.content(_:)`, which is the only place that knows it.
///
/// `.zero` by default, and that default is load-bearing: any node rendered
/// outside a `RovenuePaywallView` (a host embedding a node view directly, a
/// preview) gets it, and `isNodeOnScreen` fails open on it.
private struct PaywallViewportFrameKey: EnvironmentKey {
    static let defaultValue: CGRect = .zero
}

extension EnvironmentValues {
    var paywallViewportFrame: CGRect {
        get { self[PaywallViewportFrameKey.self] }
        set { self[PaywallViewportFrameKey.self] = newValue }
    }
}
